std/experimental/allocator/package.d

```diff
@@ -1123,48 +1123,107 @@ allocated with the same allocator.
 */
 void dispose(A, T)(auto ref A alloc, T* p)
 {
+    if (!p) return;
     static if (hasElaborateDestructor!T)
-    {
-        destroy(*p);
-    }
+        p.__dtor;
     alloc.deallocate((cast(void*)p)[0 .. T.sizeof]);
 }
 
 /// Ditto
 void dispose(A, T)(auto ref A alloc, T p)
-if (is(T == class) || is(T == interface))
+if (is(T == class) && T.stringof != Object.stringof)
+{
+    if (!p) return;
+    auto support = (cast(void*) p)[0 .. typeid(p).initializer.length];
+    static if (__traits(hasMember, T, "__dtor"))
```

Comment:
This won't work always. For example:

```
class Base {}
class Derived : Base { ~this() { writeln("derived dtor"); }

Base base = Mallocator.instance.make!Derived;
Mallocator.instance.dispose(base);
// Since Base doesn't have a dtor, Derived's dtor won't be called.
```

Furthermore, I'm not sure if the `p.__dtor` call is actually dynamically dispatched.